firedrake/matrix.py

```diff
@@ -1,13 +1,108 @@
+
+    :arg a: the bilinear form this :class:`MatrixBase` represents.
+
+    :arg bcs: an iterable of boundary conditions to apply to this
+        :class:`MatrixBase`.  May be `None` if there are no boundary
+        conditions to apply.
+    """
+    def __init__(self, a, bcs):
+        self._a = a
+
+        # Iteration over bcs must be in a parallel consistent order
+        # (so we can't use a set, since the iteration order may differ
+        # on different processes)
+        self._bcs = [bc for bc in bcs] if bcs is not None else []
+
+        test, trial = a.arguments()
+        self.comm = test.function_space().comm
+        self.block_shape = (len(test.function_space()),
+                            len(trial.function_space()))
+
+    def assemble(self):
+        raise NotImplementedError
+
+    @property
+    def assembled(self):
+        raise NotImplementedError
+
+    @property
+    def has_bcs(self):
+        """Return True if this :class:`MatrixBase` has any boundary
+        conditions attached to it."""
+        return self._bcs != []
+
+    @property
+    def bcs(self):
+        """The set of boundary conditions attached to this
+        :class:`.MatrixBase` (may be empty)."""
+        return self._bcs
+
+    @bcs.setter
+    def bcs(self, bcs):
+        """Attach some boundary conditions to this :class:`MatrixBase`.
+
+        :arg bcs: a boundary condition (of type
+            :class:`.DirichletBC`), or an iterable of boundary
+            conditions.  If bcs is None, erase all boundary conditions
+            on the :class:`.MatrixBase`.
+        """
+        self._bcs = []
+        if bcs is not None:
+            try:
+                for bc in bcs:
+                    self._bcs.append(bc)
+            except TypeError:
+                # BC instance, not iterable
+                self._bcs.append(bcs)
+
+    @property
+    def a(self):
+        """The bilinear form this :class:`.MatrixBase` was assembled from"""
+        return self._a
+
+    def add_bc(self, bc):
+        """Add a boundary condition to this :class:`MatrixBase`.
+
+        :arg bc: the :class:`.DirichletBC` to add.
+
+        If the subdomain this boundary condition is applied over is
+        the same as the subdomain of an existing boundary condition on
+        the :class:`MatrixBase`, the existing boundary condition is
+        replaced with this new one.  Otherwise, this boundary
+        condition is added to the set of boundary conditions on the
+        :class:`MatrixBase`.
+        """
+        new_bcs = [bc]
+        for existing_bc in self._bcs:
+            # New BC doesn't override existing one, so keep it.
+            if bc.sub_domain != existing_bc.sub_domain:
+                new_bcs.append(existing_bc)
+        self._bcs = new_bcs
+
+    def __repr__(self):
+        raise NotImplementedError
+
+    def __str__(self):
+        raise NotImplementedError
+
+    def force_evaluation(self):
+        raise NotImplementedError
```

Comment:
I think [abstract methods](https://docs.python.org/2/library/abc.html#abc.abstractmethod) should be preferred over having a bunch of methods raising `NotImplementedError`.